src/Compilers/CSharp/Portable/Symbols/Source/LambdaSymbol.cs

```diff
@@ -369,22 +369,17 @@ public sealed override bool Equals(Symbol symbol, TypeCompareKind compareKind)
             if ((object)this == symbol) return true;
 
             return symbol is LambdaSymbol lambda
-                && areEqual(lambda.syntaxReferenceOpt, syntaxReferenceOpt)
+                && lambda.Syntax == Syntax
                 && lambda._refKind == _refKind
                 && TypeSymbol.Equals(lambda.ReturnType, this.ReturnType, compareKind)
                 && ParameterTypesWithAnnotations.SequenceEqual(lambda.ParameterTypesWithAnnotations, compareKind,
                                                                (p1, p2, compareKind) => p1.Equals(p2, compareKind))
                 && lambda.ContainingSymbol.Equals(ContainingSymbol, compareKind);
-
-            static bool areEqual(SyntaxReference a, SyntaxReference b)
-            {
-                return (object)a.SyntaxTree == b.SyntaxTree && a.Span == b.Span;
-            }
         }
 
         public override int GetHashCode()
         {
-            return syntaxReferenceOpt.GetHashCode();
+            return Syntax.GetHashCode();
```

Comment:
>Syntax.GetHashCode(); [](http://example.com/codeflow?start=19&length=21)

I assume this is where the root cause of the problem was. SyntaxReference uses reference equality, therefore distinct syntax references almost always have different hash code. If this is correct, then I think we should simply align this implementation with the original implementation of ```Equals```. I.e. combine hash codes of the tree and span and revert changes to the ```Equals```.

Reply:
@AlekseyTs It was exactly my intention to keep the original implementation. The changes that used the tree and the span along the `GetHashCode` change were all introduced in https://github.com/dotnet/roslyn/pull/52178. This simply reverts this small portion of #52178. I'm okay with any of the suggested approaches by you or @cston

I just wanted to point out to #52178 since this is what took me this way.

Reply:
> It was exactly my intention to keep the original implementation. 

When I said "original", I meant "original" in context of this PR. And the suggestion to revert ```Equals``` also in context of changes made in this PR.

Reply:
To clarify. I think that we shouldn't change ```Equals``` in this PR and should simply align implementation of ```GetHashCode``` with what ```Equals``` is doing.

Reply:
@AlekseyTs Thanks. Fixed.

My initial thought was to keep the implementation that has been working for years, but I can also see this is not necessarily correct (ie, it may have changed for a good reason).

Though for this PR, it's likely both implementations will work equivalently anyway, so keeping `Equals` as-is and do minimal change in `GetHashCode` does make more sense.

Thanks again :)